### ISSUE TITLE
Basic WASM support on stable rust (Question)

### ISSUE THREAD

**Issue by mmulli** [Amanieu/parking_lot#310]
Hi,

After looking at the readme section titled "Nightly vs Stable" I decided to stay away from using Parking lot on WASM due to this note:

"The wasm32-unknown-unknown target is only supported on nightly and requires -C target-feature=+atomics in RUSTFLAGS."

So I invested some time into making a platform configured wrapper type that uses RefCell for the wasm target and parking lot's RwLock for others.

However, reading around a few places I've been thinking that maybe there is some basic support for WASM without nightly?

For example, in [thread parker](https://github.com/Amanieu/parking_lot/blob/master/core/src/thread_parker/mod.rs) there is this section which does not require nightly (Though I get this may panic so perhaps this isn't the most appropriate example).
```rust
    // ..
    else if #[cfg(target_family = "wasm")] {
        #[path = "wasm.rs"]
        mod imp;
    }
```

And perhaps more importantly I noted the following from [Embedonomicon](https://docs.rust-embedded.org/embedonomicon/custom-target.html#fill-the-target-file):

[If you have] a single-core processor, no threads, no interrupts, or any way for multiple things to be happening in parallel: if you are sure that is the case, such as WASM (for now), you may set "singlethread": true. **This will configure LLVM to convert all atomic operations to use their single threaded counterparts**. Incorrectly using this option may result in UB if using threads or interrupts.

In short I'm just after confirmation or a way for parking lot on stable rust + WASM to have it's use of atomics fallback to using their single threaded counterparts. If that's the case then I won't need to introduce my own wrapper type. I also think this might help with support for running parking lot on the Safari browser which I think is a bit behind in the area of experimental threading.

Hopefully this makes sense and happy new year! :)

**Comment by mmulli** [comment]
Quick follow up.  With this code:

```rust
let x = RwLock::new(1);
let x = x.try_read().unwrap();
```

On the latest Crates.io version (0.11.2), I get the following runtime error after building for WASM with nightly and running on chrome:
```
Uncaught TypeError: Failed to resolve module specifier "env". Relative references must start with either "/", "./", or "../".
```

However, on the latest master version I no longer get a runtime error and I've not added -C target-feature=+atomics. So it seems that adding those RUSTFLAGS is no longer strictly required to get some minimal level of RwLock functionality working on Wasm?

**Comment by Amanieu** [comment]
That error looks unrelated to parking_lot, maybe it's an issue with something else in your toolchain.

The comment in the README is actually somewhat misleading. parking_lot works just fine without `+atomics`, but any attempt to block (since there are no threads, this can only happen with a self-deadlock) will cause a panic. `Mutex` and other types will still work but won't do anything useful since the atomics that they use are converted by rustc to non-atomic operations. This is fine if there are no other threads sharing the same WASM memory.

**Comment by mmulli** [comment]
Thanks for clarifying. It's incredibly useful that this works to a basic extent so that more code can be shared between single threaded wasm and multi threaded native builds.
